`bh-card-list` and `bh-card` components

sfount opened the pull request:
This pull request implements the proposed `bh-card-list` API found in https://github.com/IMA-WorldHealth/bhima-2.X/issues/3135. 

This component is then applied to the Cashbox module, significantly contributing to https://github.com/IMA-WorldHealth/bhima-2.X/issues/3122. 

The `bh-card-list` component displays a list of cards provided a list of entities, it provides sorting and filtering functionality on these entities. 

**Example usage:** 
```html
<bh-card-list
  data = "arrayOfThings"
  name = "thingName"
  template = "/path/to/my/thing/template.card.html">
</bh-card-list>
```

This will generate the following (example the cashbox implementation in this PR):
![screenshot from 2018-09-04 08-20-39](https://user-images.githubusercontent.com/2844572/45018804-b95c2300-b022-11e8-8443-259fa06834e6.png)

Note a benefit of the larger card layout is making actionable information very clear to the user - in this example cashboxes that no users can access are highlighted with a red exclamation point, administrators will have to provide permissions before these can be used in the application. 

In order to significantly reduce boiler plate for implementing this design pattern in other modules, this PR also implements the `bh-card` component, this uses [multiple transclusion](https://docs.angularjs.org/api/ng/directive/ngTransclude) introduced in angular 1.5x to template in data at different points in the cards layout. 

**Example usage:** 
```html
<bh-card>
  <bh-card-title>{{card.label}}</bh-card-title>
  <bh-card-subtitle>Project: {{card.project_name}}</bh-card-sutitle>
  <bh-card-edit><a ui-sref="card.edit">Edit Entity</bh-card-edit>
  <bh-card-detail>{{card.category_label}}</bh-card-detail>
  <bh-card-statistic>{{card.number_of_users}} total users</bh-card-statistic>
</bh-card>
```

Sorting and filtering are automatically applied based on the information passed through the `name` and optional `age` and `size` bindings. Any properties that aren't bound will simply not be shown to the user. 

**Filtering**
![screenshot from 2018-09-04 08-21-35](https://user-images.githubusercontent.com/2844572/45018911-fcb69180-b022-11e8-8fac-132e1460fd47.png)

**Sorting**
![screenshot from 2018-09-04 09-14-04](https://user-images.githubusercontent.com/2844572/45018917-017b4580-b023-11e8-8334-213ffa20932f.png)



Contributes to https://github.com/IMA-WorldHealth/bhima-2.X/issues/3135.

sfount committed:
feat(bhcardlist): initial component files

This commit introduces the initial controllers and templates required
for the bh-card-list component. It will initially be developed for the
cashbox mangement module.

sfount committed:
feat(bhcardlist): filtering and ordering on list

This commit exposes optional ordering on age and size attributes. It
also implements ordering and filtering on the required name attribute.

sfount committed:
feat(bhcard): introduced bh-card component

This commit introduces the bh-card component to allow simple card
definitions to be templated into bh-card-list, it does this through
multiple transclusion introduced in angularjs 1.5.

It also integrates the cashbox module with the initial version of
bh-card-list.

sfount committed:
chore(cashbox): integrate bh-card-list into module

This commit integrates the bh-card-list and bh-card entities into the
Cashbox administration module, it also adds translation keys and cleans
up component code.

jniles commented on client/src/js/components/bhCardList.js:
```diff
@@ -0,0 +1,100 @@
+
+    // parse optional bindings
+    if (angular.isDefined(age)) {
+      $ctrl.orderOptions.available.push(
```
Just FYI, you can pass multiple parameters into `Array.prototype.push`, so you could write:
```js
$ctrl.orderOptions.available.push(
  { attribute : age, key : 'TABLE.COLUMNS.SORTING.CREATED_ASC', reverse : false },
  { attribute : age, key : 'TABLE.COLUMNS.SORTING.CREATED_DSC', reverse : true }
);
```
to achieve the same result.

jniles commented on client/src/modules/cash/cashboxes/cashboxes.js:
```diff
@@ -13,7 +13,7 @@ CashboxController.$inject = [
  * ease of use thought the application.
  */
-function CashboxController(Session, Projects, Boxes, util, Notify, $state) {
+function CashboxController(Session, Projects, CashBoxes, util, Notify, $state) {
```
Good choice of renaming 👍

jniles commented on client/src/modules/templates/bhCard.tmpl.html:
```diff
@@ -0,0 +1,23 @@
+        <h4 style="margin-top : 0px;" ng-transclude="title"></h4>
+        <h5 style="margin-top : 0px;" ng-transclude="subtitle"></h5>
+      </div>
+      <div class="col-md-6 text-float-util">
```
What is `text-float-util`?  I think this is first time I've seen it!

sfount replied:
Apparently and bizarrely it sets `text-align: right`. `text-float-util` is just a ruse to make the programmer think this is more important than it is :disappointed_relieved:. 

It looks like this is a carry over from older code, I will _consider_ using the more aptly named bootstrap `text-right`.

jniles commented on client/src/modules/templates/bhCardList.tmpl.html:
```diff
@@ -0,0 +1,45 @@
+  ng-repeat="card in $ctrl.data
+    | filter : $ctrl.filterOptions.value : strict
+    | orderBy : 'label':$ctrl.orderOptions.current.reverse">
+  <ng-include src="$ctrl.template"></ng-include>
```
Should we support pagination?  I know our Fiscal Year controller already has support for it via `uib-pagination`.  Do you think we could put it in here?

sfount replied:
I did notice pagination on the fiscal year controller. I think with all the modules in line to support this pattern none of them will have an overwhelming number of entities that would require both filtering and paginating. 

Definitely something to consider for future improvement though! The nice thing about setting up like this will be updating the Component to support pagination across all modules that implement this.

jniles commented on client/src/modules/templates/bhCardList.tmpl.html:
```diff
@@ -0,0 +1,45 @@
+    ng-model="$ctrl.filterOptions.value[$ctrl.currentFilterAttribute]"
+    ng-show="$ctrl.filterOptions.active"
+    class="form-control"
+    placeholder="{{'FORM.PLACEHOLDERS.FILTER_NAME' | translate }}"
```
Just FYI - with any plain HTML element, you can use `angular-translate` attributes for a slight performance increase, instead of using the expensive pipe.  However, this does not work for custom components (read: `ui-select`), which is why you still see this pattern in those components.

To use the translate attributes, you can use the [translate-attr](https://angular-translate.github.io/docs/#/api/pascalprecht.translate.directive:translate-attr) directive to do it in one pass:
```html
<input
  translate-attr="{ placeholder : 'FORM.PLACEHOLDERS.FILTER_NAME' }">
```

sfount replied:
Good to know!

jniles commented on server/controllers/finance/cashboxes/index.js:
```diff
@@ -36,9 +36,11 @@ exports.privileges = privileges;
-    'SELECT id, label, is_auxiliary, project_id FROM cash_box ';
+  let sql = 'SELECT id, label, is_auxiliary, project_id FROM cash_box ';
 
+  // @TODO(sfount) this isn't a detailed query for cashboxes because it returns
```
Yeah, good call!  We can change the API for this down the road.

jniles reviewed (commented):
@sfount, great work on this.  It's a very nice prototype already!

I'd be interested in writing some unit tests for this code before merging. It will serve as a really good example of how to unit test logic (sorting, filtering) in both the controller and the DOM.  If I write the tests, it would also go to validate your API design (unless it seems super confusing, in which case, it would prompt reforms).  What do you think?  Shall I make an attempt or do you think it isn't stabilized yet?

sfount committed:
chore(cashbox): apply standards to card list tmpl.

This commit updates the card list template and controller to ensure code
is legible and the template uses the latest repository standards. It
also resolve rebase conflicts from upstream.

jniles committed:
test(bhCardList): add unit tests for card lists

This commit add client unit tests for the bhCardList component, testing
all portions of the code.  There seems to be some issues with filtering
tests, so the filtering test is skipped.  It may just be my lack of
knowledge about the API though.

sfount commented on test/client-unit/components/bhCardList.spec.js:
```diff
@@ -0,0 +1,175 @@
+    expect(filterInput).to.have.class('ng-hide');
+  });
+
+  it.skip('should support filtering the values in the DOM', () => {
```
It looks like this is behaving as it should! I haven't tried this yet but you're setting the filter value (name, as you have well suggested changing) to be `title`. 

It looks like all three titles contain the word 'Card' and we are then filtering on 'd' so all 3 should be rendered. If the `name` attribute was set to `text` I think the results would be as expected, with only 1 rendered.

sfount replied:
To update this: It looks like the offending line is on the input:
```html
ng-model = "$ctrl.filterOptions[$ctrl.name]"
```

In the browser this correctly evaluates as follows: 
![screenshot from 2018-09-10 16-37-32](https://user-images.githubusercontent.com/2844572/45308068-8fba6480-b518-11e8-9c0b-86ba1bc5d7db.png)

But in the client test runner the $ctrl.name is not respected: 
![screenshot from 2018-09-10 16-37-15](https://user-images.githubusercontent.com/2844572/45308093-9cd75380-b518-11e8-8f7a-6ee424976303.png)

It looks like a quirk with the test runner but I will look into it a bit more.

sfount reviewed (commented):
These are really great unit tests, very clear and easy to read! The skipped test shouldn't cause too much issue. 

I'm not sure why the yarn.lock file was included? Would you like to fix this test or should I pull it, rebase and then push a version to be merged?

EDIT: I agree that the name change suggestions you made should be applied, they can be made along with the last test and resolving yarn lock issues.

sfount committed:
chore(bhCardList): rename component API; fix filtering issue

This commit renames the exposed binding on the bh-card-list component
to more clearly represent their purpose, favouring sort-name, sort-age
and sort-size over name, age and size respectively.

It also resolves an issue with unit tests failing on filtering due to an
issue with ng-model with a variable attribute, simply supporting
filtering all elements on all attributes resolved this issue.

jniles committed:
chore(i18n): add translations

This commit adds translations for the cashbox keys.
